src/components/toasts-list/components/transaction-toast/components/transaction-toast-progress/transaction-toast-progress.tsx

```diff
@@ -1,31 +1,85 @@
-    if (!this.currentRemaining) {
-      return <slot></slot>;
+  private DEFAULT_INFINITE_ANIMATION_DURATION = 30;
+
+  @State() currentTimestamp: number = Date.now() / 1000;
+  @State() hasTimeElapsed: boolean = false;
+  @State() expectedTransactionDuration: number = 0;
+  @State() secondsPassedSinceStart: number = 0;
+  @State() shouldShowProgressBar: boolean = false;
+  @State() percentagePassedSinceStart: number = 0;
+  @State() shouldQuickFill: boolean = false;
+  @State() infiniteProgressDelay: number = 0;
+  @State() infinitePercentagePassedSinceStart: number = 0;
+  @State() infinitePercentageAnimationDuration: number = this.DEFAULT_INFINITE_ANIMATION_DURATION + (this.endTime - this.startTime) * 2;
+
+  componentWillLoad() {
+    console.log(this.endTime - this.startTime);
+    this.updateProgress();
+  }
+
+  @Watch('startTime')
+  @Watch('endTime')
+  handleTimeChange() {
+    this.updateProgress();
+  }
+
+  private updateProgress() {
+    this.shouldShowProgressBar = Boolean(this.endTime) && Boolean(this.startTime);
+
+    if (!this.shouldShowProgressBar) {
+      this.shouldQuickFill = true;
+      setTimeout(() => (this.hasTimeElapsed = true), 500);
+      return;
     }
 
+    this.currentTimestamp = Date.now() / 1000;
+    this.expectedTransactionDuration = this.endTime - this.startTime;
+    this.secondsPassedSinceStart = this.currentTimestamp - this.startTime;
+    this.percentagePassedSinceStart = this.expectedTransactionDuration > 0 ? (this.secondsPassedSinceStart / this.expectedTransactionDuration) * 100 : 0;
+    this.infiniteProgressDelay = Math.max(0, this.expectedTransactionDuration - this.secondsPassedSinceStart);
+    this.infinitePercentagePassedSinceStart = (this.secondsPassedSinceStart / (this.expectedTransactionDuration + this.infinitePercentageAnimationDuration)) * 100;
+  }
+
+  render() {
     return (
-      <div class={this.progressClass}>
-        <div
-          class="transaction-toast-bar"
-          style={{ width: `${this.currentRemaining}%` }}
-          role="progressbar"
-          aria-valuemin={0}
-          aria-valuemax={100}
-          aria-valnow={this.currentRemaining}
-        />
-        <slot></slot>
-      </div>
+      <Fragment>
+        <div class="transaction-toast-bar-wrapper" style={{ opacity: this.hasTimeElapsed ? '0' : '1' }}>
+          <div
+            class="transaction-toast-bar fixed"
+            style={{
+              '--animation-duration': `${this.expectedTransactionDuration}s`,
+              '--start-width': `${this.percentagePassedSinceStart}%`,
+            }}
+          />
+
+          <div class="transaction-toast-bar infinite">
+            <div
+              class="transaction-toast-bar-line"
+              style={{
+                '--start-width': `${this.infinitePercentagePassedSinceStart}%`,
+                '--animation-duration': `${this.infinitePercentageAnimationDuration}s`,
+                '--animation-delay': `${this.infiniteProgressDelay}s`,
```

Comment:
Where do these CSS variables come frome?

Reply:
They don't come from anywhere. They're being declared here and passed on in CSS.